### PR TITLE
boost: add Boost.URL

### LINKED ISSUE #14991: [package] boost/1.81: new URL library is not included in the build
### Description

Boost 1.81 introduces the new URL library: https://www.boost.org/doc/libs/1_81_0/libs/url/doc/html/index.html.

It is not included in the boost recipe.

Related PR ##14888.

### Package and Environment Details

* Package Name/Version: **boost/1.81**
* Operating System+version: Any
* Compiler+version: Any
* Docker image: Any
* Conan version: **conan 1.55.0**
* Python version: **Python 3.7.4**

### Conan profile

[settings]
os=Macos
os_build=Macos
arch=armv8
arch_build=armv8
compiler=apple-clang
compiler.version=14
compiler.libcxx=libc++
build_type=Release
[options]
[conf]
[build_requires]
[env]

### Steps to reproduce

conan install boost/1.81

### Logs

<details><summary>Click to expand log</summary>

```
Undefined symbols for architecture x86_64:
  "boost::urls::parse_uri(boost::core::basic_string_view<char>)", referenced from:
      _main in main.cpp.o
  "boost::urls::url_view_base::encoded_host() const", referenced from:
      boost::urls::string_token::return_string::result_type boost::urls::url_view_base::host<boost::urls::string_token::return_string>(boost::urls::string_token::return_string&&) const in main.cpp.o
  "boost::urls::url_view_base::port() const", referenced from:
      _main in main.cpp.o
  "boost::urls::url_view_base::scheme() const", referenced from:
      _main in main.cpp.o
  "boost::urls::pct_string_view::decode_impl(boost::urls::string_token::arg&, boost::urls::encoding_opts) const", referenced from:
      boost::urls::string_token::return_string::result_type boost::urls::pct_string_view::decode<boost::urls::string_token::return_string>(boost::urls::encoding_opts, boost::urls::string_token::return_string&&) const in main.cpp.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [srt-rendezvous] Error 1
```

</details>

### PR DESCRIPTION
**Pull request by sanblch**
Specify library name and version:  **boost/1.81.0**
Closes #14991

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by prince-chrismc** [comment]
Did you run the rebuild dependencies script by any chance?

**Comment by sanblch** [comment]
@prince-chrismc, thank you. I did check and made fixes accordingly. The script added system lib dependency to the URL lib.

**Comment by sanblch** [comment]
> Did you run the rebuild dependencies script by any chance?

@prince-chrismc, thank you. I did check and made fixes accordingly. The script added system lib dependency to the URL lib.

**Review by uilianries — DISMISSED** [review]

**Review by jwillikers — DISMISSED** [review]

**Review by SSE4 — DISMISSED** [review]

**Review by prince-chrismc — DISMISSED** [review]

**Comment by prince-chrismc** [comment]
There's merge conflicts now

**Comment by prince-chrismc** [comment]
Also if you could be so kind to sneak in https://github.com/conan-io/conan-center-index/pull/15105#issuecomment-1372799474 id appreciate it ❤️

File: recipes/boost/all/test_package/CMakeLists.txt [hunk 1]
```diff
@@ -116,6 +116,14 @@ if(NOT HEADER_ONLY)
         target_include_directories(numpy_exe PRIVATE ${PYTHON_INCLUDE_DIRS})
         target_link_libraries(numpy_exe PRIVATE Boost::numpy${PYTHON_COMPONENT_SUFFIX} ${PYTHON_LIBRARIES})
     endif()
+
+    if(WITH_URL)
+        find_package(Boost COMPONENTS url REQUIRED)
+        add_executable(url_exe url.cpp)
+        target_link_libraries(url_exe PRIVATE Boost::url)
+        set_property(TARGET url_exe PROPERTY CXX_STANDARD 11)
```
Comment (jcar87) [inline recipes/boost/all/test_package/CMakeLists.txt:124]: once rebasted on top of the most recent revision of the upstream `master` branch, this now needs:

```suggestion
        set_property(TARGET url_exe PROPERTY CXX_STANDARD 11)
        add_test(NAME boost_url COMMAND url_exe)
```

File: recipes/boost/all/test_package/conanfile.py [hunk 2]
```diff
@@ -46,6 +46,7 @@ def generate(self):
         tc.variables["WITH_STACKTRACE"] = not self.dependencies["boost"].options.without_stacktrace
         tc.variables["WITH_STACKTRACE_ADDR2LINE"] = self.deps_user_info["boost"].stacktrace_addr2line_available
         tc.variables["WITH_STACKTRACE_BACKTRACE"] = self._boost_option("with_stacktrace_backtrace", False)
+        tc.variables["WITH_URL"] = not self._boost_option("without_url", True)
```
Comment (jcar87) [inline recipes/boost/all/test_package/conanfile.py:49]: once rebased, this needs
```suggestion
        tc.cache_variables["WITH_URL"] = not self._boost_option("without_url", True)
```

good catch noticing that this needs to evaluate the option via `_boost_option` :)

File: recipes/boost/all/test_package/conanfile.py [hunk 3]
```diff
@@ -99,3 +100,5 @@ def test(self):
                 self.run(os.path.join(bindir, "stacktrace_basic_exe"), env="conanrun")
             if self._boost_option("with_stacktrace_backtrace", False):
                 self.run(os.path.join(bindir, "stacktrace_backtrace_exe"), env="conanrun")
+        if not self._boost_option("without_url", True):
+            self.run(os.path.join(bindir, "url_exe"), env="conanrun")
```
Comment (jcar87) [inline recipes/boost/all/test_package/conanfile.py:104]: Once rebased, the logic in this function can be omitted

```suggestion
```

File: recipes/boost/all/test_v1_package/conanfile.py [hunk 4]
```diff
@@ -87,3 +88,5 @@ def test(self):
                 self.run(os.path.join("bin", "stacktrace_basic_exe"), run_environment=True)
             if self._boost_option("with_stacktrace_backtrace", False):
                 self.run(os.path.join("bin", "stacktrace_backtrace_exe"), run_environment=True)
+        if not self._boost_option("without_url", True):
+            self.run(os.path.join("bin", "url_exe"), run_environment=True)
```
Comment (jcar87) [inline recipes/boost/all/test_v1_package/conanfile.py:92]:
```suggestion
```

once rebased, this logic can go entirely (this block doesnt exist in the current head)

**Review by jcar87 — CHANGES_REQUESTED** [review]
HI @sanblch, thank you so much for your contribution.

I'm afraid this recipe has been updated in our master branch and this PR is now a little out of sync and would require some minor changes - I've added suggestions to apply once your downstream branch is synced up with the most recent changes.

File: recipes/boost/all/test_package/conanfile.py (hunk 2 above)
Comment (sanblch) [reply recipes/boost/all/test_package/conanfile.py]: Thank you. Done.

File: recipes/boost/all/test_package/conanfile.py (hunk 3 above)
Comment (sanblch) [reply recipes/boost/all/test_package/conanfile.py]: Done.

File: recipes/boost/all/test_v1_package/conanfile.py (hunk 4 above)
Comment (sanblch) [reply recipes/boost/all/test_v1_package/conanfile.py]: Done.

**Comment by sanblch** [comment]
> HI @sanblch, thank you so much for your contribution.
> 
> I'm afraid this recipe has been updated in our master branch and this PR is now a little out of sync and would require some minor changes - I've added suggestions to apply once your downstream branch is synced up with the most recent changes.

HI @jcar87, thank you for your help! Test checks became way better.

**Comment by prince-chrismc** [comment]
Please do not force push 🙏 GitHub forces us to restart the review which is not fun!

**Review by jwillikers — APPROVED** [review]

**Comment by sanblch** [comment]
> Please do not force push 🙏 GitHub forces us to restart the review which is not fun!

I'm sorry for that. I reread the guides and will change my habits.

**Review by prince-chrismc — APPROVED** [review]

**Comment by prince-chrismc** [comment]
@jcar87 you need to follow up since you requested changes :)

**Comment by uilianries** [comment]
@jcar87 a friendly ping. This PR which is blocked and need your review to unblock 😸

**Comment by prince-chrismc** [comment]
Pinged him on slack for better results

**Review by jcar87 — APPROVED** [review]
LGTM! Thanks @sanblch

**Comment by prince-chrismc** [comment]
Thank you 😊 💓 ☺️ 💗 💛 💖 😊 💓 ☺️ 💗 💛 💖 😊

File: recipes/boost/all/conanfile.py [hunk 5]
```diff
@@ -434,7 +438,7 @@ def layout(self):
 
     @property
     def _cxx11_boost_libraries(self):
-        libraries = ["fiber", "json", "nowide"]
+        libraries = ["fiber", "json", "nowide", "url"]
         if Version(self.version) >= "1.76.0":
             libraries.append("math")
         if Version(self.version) >= "1.79.0":
```
Comment (maxsharabayko) [inline recipes/boost/all/conanfile.py:444]: Shouldn't the `"url"` go under the `if Version(self.version) >= "1.81.'":`?

Comment (sanblch) [reply recipes/boost/all/conanfile.py:444]: That would also be correct. But this is a set of c++11 libraries. Libraries that not always c++11 are applied by version. From that point "url" is always c++11. Its appearance in 1.81 is set by `dependencies/dependencies-1.81.0.yml`.

**Pull request merged**